### PR TITLE
nixos/qemu-vm: Ensure 9pnet_virtio module is loaded for shared dirs

### LINKED ISSUE #240175: nixos/systemd: A volatile `/` (i.e. on a tmpfs) does not work with the systemd initrd and 9p mounts
### Describe the bug

The systemd initrd (or stage1) does not work with a volatile root when there are 9p mounts present. 

A result of this is that VMs built with nixos/qemu-vm do not work when disabling the `diskImage` (and thus running `/` on a tmpfs) and using the systemd initrd.

### Steps To Reproduce

Create a test that sets these options:

```nix
virtualisation.diskImage = null;
boot.initrd.systemd.enable = true;
```

### Expected behavior

The test should work just like the test works in https://github.com/NixOS/nixpkgs/pull/238848

### Notify maintainers

@RaitoBezarius 
@lilyinstarlight who has identified the root cause (an ordering issue where the tmpfs is mounted before the 9p infrastructure is up) and found a hack around this behaviour.

### PR DESCRIPTION
**Pull request by ElvishJerricco**
## Description of changes

Fixes #240175

When `diskImage = null`, the root fs is a tmpfs instead of `/dev/vda`. Thus, it doesn't have to wait for virtio modules to load before being mounted. The root fs is a dependency of shared directories by nature of being their parent directory. Without depending on `/dev/vda`, these shared directories may attempt to mount without virtio modules being loaded.

/cc @nikstur 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by nikstur** [comment]
In my mental model, the kernel was supposed to load these modules on demand when it tries to mount a 9P share, but I guess it doesn't. Did you find out anything about that?

**Comment by ElvishJerricco** [comment]
@nikstur Yes, but these are not dependencies of the `9p` module.

```
$ modinfo 9p
....
depends:        9pnet,netfs
...

$ modinfo 9pnet
...
depends:        netfs
...

$ modinfo 9pnet_virtio
...
depends:        virtio,9pnet,virtio_ring
...
```

So just trying to mount the file system won't load the virtio drivers.

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -1045,7 +1045,8 @@ in
 
     boot.initrd.availableKernelModules =
       optional (cfg.qemu.diskInterface == "scsi") "sym53c8xx"
-      ++ optional (cfg.tpm.enable) "tpm_tis";
+      ++ optional (cfg.tpm.enable) "tpm_tis"
+      ++ optional  (cfg.sharedDirectories != {}) "9pnet_virtio";
```
Comment (nikstur) [inline nixos/modules/virtualisation/qemu-vm.nix:1049]: We already do this in the guest profile: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/profiles/qemu-guest.nix

**Review by nikstur — APPROVED** [review]
Looks good. We should just remove the duplicate 9pnet_virtio module

File: nixos/modules/virtualisation/qemu-vm.nix (hunk 1 above)
Comment (ElvishJerricco) [reply nixos/modules/virtualisation/qemu-vm.nix:1049]: Ah, fair point. I'll fix, test, and merge.

**Pull request merged**